Assets/BossRoom/Scripts/Server/Game/Action/DashAttackAction.cs

```diff
@@ -0,0 +1,127 @@
+using MLAPI;
+using MLAPI.Spawning;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace BossRoom.Server
+{
+    /// <summary>
+    /// Causes the attacker to dash toward a target. When they stop dashing, they perform a melee attack.
+    /// The character is immune to damage while dashing.
+    ///
+    /// Since the "Range" field means "range when we can begin dashing at our target", we need another
+    /// field to mean "range of our melee attack after dashing". We'll use the "Radius" field of the
+    /// ActionDescription for that.
+    /// 
+    /// </summary>
+    /// <remarks>
+    /// See MeleeAction for relevant discussion about targeting; we use the same concept here: preferring
+    /// the chosen target, but using whatever is actually within striking distance at time of attack.
+    /// </remarks>
+    /// 
+    public class DashAttackAction : Action
+    {
+        private bool m_StartedDash;
+        private bool m_StoppedDash;
+        private float m_DashDuration;
+
+        public DashAttackAction(ServerCharacter parent, ref ActionRequestData data) : base(parent, ref data)
+        {
+            Assert.IsTrue(Description.MoveSpeed > 0, $"ActionDescription for {Description.ActionTypeEnum} needs a MoveSpeed assigned!");
+            Assert.IsTrue(Description.Radius > 0, $"ActionDescription for {Description.ActionTypeEnum} needs a Radius assigned!");
+        }
+
+        public override bool Start()
+        {
+            // snap to face our destination. This is the direction we'll charge in
+            m_Parent.transform.LookAt(GetTargetSpot());
+
+            // tell clients to visualize this action
+            m_Parent.NetState.RecvDoActionClientRPC(Data);
+            return true;
+        }
+
+        public override bool Update()
+        {
+            if (!m_StartedDash && TimeRunning >= Description.ExecTimeSeconds)
```

Comment:
This should be an instant action. An RPC is sent to the client. This is in opposition to the boss charge which moves your character over time, here an RPC that can't be corrected teleports your player.

Reply:
That isn't what the art is set up for, and it's not what's is in the GDD. Since reimplementing it as a teleport doesn't improve its gameplay behavior in any way I can figure, I assume you want this rewritten in order to have an example of a networked teleport in the demo?

I'm happy to do so, but... I'm not sure how it's supposed to work. To avoid further delays I'd like to get answers to some design questions first! (Paging @dwoodruffsf, but figured I'd do this conversation here in the PR, since I assume @SamuelBellomo has expectations also.)

1- The artwork is designed to show the character dashing across the screen. Do we need new artwork, or do we show the character "pretend running" across the screen? (Presumably by having the ActionFX disable normal visualization of networked movement, and determining the visualization's location by lerping between start and end spot.)
2- Is this a fixed-duration teleport, or should we calculate the distance based on a speed variable, as we do with mage-bolt? (Fixed duration is much simpler in terms of implementing the visualization part.)
3- Can the rogue teleport through walls? (If not, presumably we implement the same LOS check as mage-bolt, and the character abruptly stops at the wall)
4- What does a new player see if they log in right as the action is happening? (Moving the visualization via an Action will mean that if a character joins right after a rogue begins, they will see the rogue standing in the old spot, before instantly blipping to the new spot. This is an unusual edge-case and I'd assume it's fine to ignore it, but want to confirm.)
5- Is it fine to teleport to the exact spot that the target is in, or do we need to try to figure out a spot "beside" the target? (E.g. teleporting to 1 meter away from the spot.) I can see both scenarios making sense: if you to teleport to the floor switch, for instance, you don't want to teleport to "near" the floor switch. But teleporting on top of the boss would be visually confusing.

Reply:
Final thought: if we end up wanting to use new artwork for this, can I suggest we give the teleport to the mage instead? (Since they're down an action anyway.) 

Implementing a "pure teleport" would be a simpler demonstration of teleportation: presumably it would let you teleport through walls (like in Diablo 2), and it wouldn't need to do all the hacky "pretend to move the visualization" stuff; the character could just perform an animation at the start spot, then blink to the new spot after the RPC.

Reply:
After internal communication, this is my understanding of what's desired:

1- we will show the character dash across the screen (lerping position)
2- fixed duration seems fine (so all dashes will take e.g. 1.5 secs)
3- ??? (will assume rogue cannot teleport through walls, for safety, as removing it is easier than adding it)
4- ignore edge case of players joining game in the 1.5secs that a teleport is happening
5- we will teleport to "near" the spot, not "on top of" the spot.

If Unity folks want alterations to this plan, please let me know ASAP! Will begin implementation of this tomorrow

Reply:
Okay, I've revised this to be a teleport followed by a melee attack at the destination spot. (And the aforementioned client-side visualization of the character dashing.)

Using a teleport instead of a dash has some problematic gameplay behavior: prior to teleporting, the enemies will continue to perceive the character as being in their original spot. So if the "dash" part takes a long time, the imps seem to not understand where the character really is.

To minimize the impact of this, I've made the "dash" part extremely short, so that you zip across the screen in 0.7 seconds. During this 0.7 seconds the character is immune to damage, which also minimizes chances for weirdness. Prior to dashing, though, the character performs a wind-up animation for 1.3 seconds, during which they are NOT immune to damage.

I have not attempted to tweak the action's particles to match these new timings, so they may appear to be off by a few tenths of a second. But we can tweak those when the action's times are finalized.